python/ray/profiling.py

```diff
@@ -0,0 +1,270 @@
+    else:
+        return RayLogSpanRaylet(
+            worker.profiler, event_type, extra_data=extra_data)
+
+
+class Profiler(object):
+    """A class that holds the profiling states.
+
+    Attributes:
+        worker: the worker to profile.
+        events: the buffer of events.
+        lock: the lock to protect access of events.
+    """
+
+    def __init__(self, worker):
+        self.worker = worker
+        self.events = []
+        self.lock = threading.Lock()
+
+    def start_flush_thread(self):
+        t = threading.Thread(target=self._periodically_flush_profile_events)
+        # Making the thread a daemon causes it to exit when the main thread
+        # exits.
+        t.daemon = True
+        t.start()
+
+    def _periodically_flush_profile_events(self):
+        """Drivers run this as a thread to flush profile data in the
+        background."""
+        # Note(rkn): This is run on a background thread in the driver. It uses
+        # the local scheduler client. This should be ok because it doesn't read
+        # from the local scheduler client and we have the GIL here. However,
+        # if either of those things changes, then we could run into issues.
+        try:
+            while True:
+                time.sleep(1)
+                self.flush_profile_data()
+        except AttributeError:
+            # This is to suppress errors that occur at shutdown.
+            pass
+
+    # TODO(rkn): Support calling this function in the middle of a task, and
+    # also call this periodically in the background from the driver.
+    def flush_profile_data(self):
+        """Push the logged profiling data to the global control store.
+
+        By default, profiling information for a given task won't appear in the
+        timeline until after the task has completed. For very long-running
+        tasks, we may want profiling information to appear more quickly.
+        In such cases, this function can be called. Note that as an
+        aalternative, we could start thread in the background on workers that
+        calls this automatically.
+        """
+        with self.lock:
+            events = self.events
+            self.events = []
+
+        if not self.worker.use_raylet:
+            event_log_key = b"event_log:" + self.worker.worker_id
+            event_log_value = json.dumps(events)
+            self.worker.local_scheduler_client.log_event(
+                event_log_key, event_log_value, time.time())
+        else:
+            if self.worker.mode == ray.WORKER_MODE:
+                component_type = "worker"
+            else:
+                component_type = "driver"
+
+            self.worker.local_scheduler_client.push_profile_events(
+                component_type, ray.ObjectID(self.worker.worker_id),
+                self.worker.node_ip_address, events)
+
+    def add_event(self, event):
+        with self.lock:
+            self.events.append(event)
+
+
+class RayLogSpan(object):
```

Comment:
RayLogSpanNonRaylet

Reply:
I just keep the original name, but I think it's good to make the name more explicit